### PR TITLE
HTTPS using docker infrastructure

### PR DESCRIPTION
**Pull request by hkolvenbach**
This PR adds the following:
- enable HTTPS support for everything (che server, keycloak, wsagents, terminal, exec agent) if `CHE_HOST_PROTOCOL=https`
- select a certificate provider through `CHE_HTTPS_CERTIFICATE_PROVIDER`

CHE_HTTPS_CERTIFICATE_PROVIDER has 3 options:
- `file`: uses `che.pem` and `che.crt` from `instance/config/traefik/`
~~-  `letsencrypt` and `letsencrypt-staging` (default): automatically generate certificates for all endpoints through letsencrypt~~

The `file` mode makes most sense with a wildcard certificate and `SINGLE_PORT_MODE` and `CHE_SINGLEPORT_WILDCARD__DOMAIN_IPLESS=true`
(a working set of ENV variables would be `-e CHE_SINGLEPORT_WILDCARD__DOMAIN_IPLESS=true -e CHE_SINGLEPORT_WILDCARD__DOMAIN_HOST=domain.tld -e CHE_HOST=ide.domain.tld  -e CHE_HOST_PROTOCOL=https -e CHE_HTTPS_CERTIFICATE_PROVIDER=file -e CHE_MULTIUSER=true -e CHE_PORT=443 -e CHE_SINGLE_PORT=true`)

~~The `letsencrypt` part is still work in progress and is not really feasible without ACMEv2 and wildcard support, because of letsencrypts rate limits. Wildcard support was just added https://community.letsencrypt.org/t/acme-v2-and-wildcard-certificate-support-is-live/55579 so hopefully traefik will support it soon as well.~~ Won't be part of this PR.

I would like to get some feedback on the implementation, especially if its robust enough and supports all potential paths or if there might be some edge cases that will break.

### What does this PR do?
If Che is running in _Single Port_ and _Multi-User Mode_, a HTTPS certificate can be provided as a set of `che.key` and `che.crt` file. To enable HTTPS support, the following steps need to be taken:
* enable _Single Port_ and _Multi-User Mode_ as described above
* enable an IP-less wildard domain as above
* copy `che.key` (private key) and `che.crt` (full chain) certificates for your `*.domain.tld` to `instance/config/traefik/` within your Che installation
* set the following configuration either as ENV variables or in `che.env`: `CHE_HTTPS_CERTIFICATE_PROVIDER=file`  

A complete configuration therefore needs to look like this:  
```shell
CHE_SINGLE_PORT=true
CHE_SINGLEPORT_WILDCARD__DOMAIN_IPLESS=true
CHE_SINGLEPORT_WILDCARD__DOMAIN_HOST=domain.tld
CHE_HOST=ide.domain.tld
CHE_HOST_PROTOCOL=https
CHE_HTTPS_CERTIFICATE_PROVIDER=file
CHE_MULTIUSER=true
CHE_PORT=443
```

### CODE REVIEW

File: assembly/assembly-main/src/assembly/tomcat/conf/server.xml
Comment (hkolvenbach) [inline assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: @skabashnyuk I am a bit confused by the tomcat docs and not sure if the old version does the same thing. do you know which headers are needed for proper reverse proxying?

Comment (skabashnyuk) [reply assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: >do you know which headers are needed for proper reverse proxying

I think ```X-Forwarded-Proto```

Question: why did you change server.xml?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (skabashnyuk) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java]: remove if not needed

File: assembly/assembly-main/src/assembly/tomcat/conf/server.xml
Comment (skabashnyuk) [reply assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: I think value ```internalProxies``` is mandatory and has to be added back.

**Comment by skabashnyuk** [comment]
In case if this pr needed changes in server.xml it has to be tested on OpenShift and Kubernetes aswell.

File: assembly/assembly-main/src/assembly/tomcat/conf/server.xml
Comment (hkolvenbach) [reply assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: @skabashnyuk thanks for the feedback, i will test again with the old values and see if it still works.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (riuvshin) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: please add new line

File: dockerfiles/base/scripts/base/startup_04_pre_cli_init.sh
Comment (riuvshin) [inline dockerfiles/base/scripts/base/startup_04_pre_cli_init.sh]: please fix indentation

File: dockerfiles/init/manifests/che.env
Comment (riuvshin) [inline dockerfiles/init/manifests/che.env:197]: maybe we should have CHE_HOST_WS_PROTOCOL? 
we can automatically set CHE_HOST_WS_PROTOCOL="wss" if CHE_HOST_PROTOCOL="https"

File: dockerfiles/base/scripts/base/startup_04_pre_cli_init.sh
Comment (hkolvenbach) [reply dockerfiles/base/scripts/base/startup_04_pre_cli_init.sh]: done

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (hkolvenbach) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: done

File: dockerfiles/init/manifests/che.env
Comment (hkolvenbach) [reply dockerfiles/init/manifests/che.env:197]: Right now, I will detect `CHE_HOST_PROTOCOL=https` and enable `wss` accordingly. Is there any use case for having `https` with a different setting then `wss`? It should be easy to add `CHE_HOST_WS_PROTOCOL` though if we want to decouple the protocols.

**Comment by skabashnyuk** [comment]
I guess we need some docs pr too. @eivantsov wdyt?

**Comment by slxcloudtestuser2** [comment]
@garagatyi i still need to look into the ACME / letsencrypt implementation, but didn't find time for that yet. the general implementation will not change anymore though.

**Comment by hkolvenbach** [comment]
@garagatyi no problem. i still need to finish the ACME / letsencrypt part, but didn't have time for that yet. the general implementation will not change anymore though and we are already using it for the `CHE_HTTPS_CERTIFICATE_PROVIDER=file` setup

**Review by garagatyi — COMMENTED** [review]
I'm OK with the solution in general. Will be able to review it again after PR gets unwipped. 
@skabashnyuk @riuvshin @benoitf @mshaposhnik Are you ok with the solution?

**Comment by skabashnyuk** [comment]
ok

**Comment by evaryont** [comment]
@hkolvenbach is this ready to be merged with just `file` support? I'd love to see this done, as I'm running into the same problem myself.

File: assembly/assembly-main/src/assembly/tomcat/conf/server.xml
Comment (skabashnyuk) [inline assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: I see no changes here. can you revert changes in this file?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (skabashnyuk) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java]: use variables instead if string concatenating.

**Review by skabashnyuk — APPROVED** [review]

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/labels/SinglePortLabelsProvisioner.java
Comment (mshaposhnik) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/provisioner/labels/SinglePortLabelsProvisioner.java]: Remove that outdated lines

**Review by mshaposhnik — APPROVED** [review]

**Comment by hkolvenbach** [comment]
@evaryont currently updating with the latest master. merging it just with file support sounds like a reasonable idea! i will look into it

File: assembly/assembly-main/src/assembly/tomcat/conf/server.xml
Comment (skabashnyuk) [inline assembly/assembly-main/src/assembly/tomcat/conf/server.xml]: not sure I understand what was changed here. Can you explain?

File: build.include
Comment (skabashnyuk) [inline build.include]: why you need this file in a root folder?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (skabashnyuk) [reply infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java]: Log.debug("Message {} {} {}", var1, var2, var3)
https://www.slf4j.org/faq.html#logging_performance

File: build.include
Comment (hkolvenbach) [reply build.include]: sorry, included by accident (used for testing)

**Comment by skabashnyuk** [comment]
@hkolvenbach can you submit documentation pr alongside with current?

**Comment by skabashnyuk** [comment]
ci-test

File: dockerfiles/init/modules/compose/templates/docker-compose.yml.erb
Comment (garagatyi) [inline dockerfiles/init/modules/compose/templates/docker-compose.yml.erb:103]: Do we need this commented line?

**Comment by hkolvenbach** [comment]
@eivantsov unfortunately, i cant merge this PR due to the ip-validation failure. I just checked again, both used emails have a valid and signed Eclipse Contributor Agreement.

regarding the ci-test: am i correct that this failed due to a jenkins related error, not the build itself?

**Comment by hkolvenbach** [comment]
regarding the failed ip validation: should i squash and rebase it on master and then sign off again?

File: dockerfiles/init/modules/keycloak/templates/che-realm.json.erb
Comment (garagatyi) [inline dockerfiles/init/modules/keycloak/templates/che-realm.json.erb:61]: Is there a chance that this addition will break deployment to openshift [here](https://github.com/eclipse/che/blob/master/dockerfiles/keycloak/kc_realm_user.sh#L23)? @eivantsov WDYT?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerMachineCreator.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerMachineCreator.java:67]: Can you elaborate why we need to do that?

**Comment by skabashnyuk** [comment]
>regarding the failed ip validation: should i squash and rebase it on master and then sign off again?

it would be nice to do that.

> about tests . @Ohrimenko1988 is checking

**Comment by garagatyi** [comment]
@hkolvenbach please, do not push with force to the branch. It would complicate review process for maintainers.
We know about issues with the IP-validation and we will work around that after the review process is finished.

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java:57]: This parameter is not used in the injection process, so let's not make it more complex by adding an annotation.

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java:174]: We have only one usage of a method with name `makeUrl` in the class. We don't need 2 methods for that.

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/ServersMapper.java:188]: Looks like this is debugging code. Can we remove it?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortUrlRewriter.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortUrlRewriter.java:65]: In case `che.port` is 80 or 443 would this URL contain an explicit port number?

File: infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortUrlRewriter.java
Comment (garagatyi) [inline infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortUrlRewriter.java:61]: can you add unit tests for the new cases?

File: dockerfiles/init/modules/keycloak/templates/che-realm.json.erb
Comment (ghost) [reply dockerfiles/init/modules/keycloak/templates/che-realm.json.erb:61]: It will, since currently I sed it run to replace puppet stuff with an actual URL - https://github.com/eclipse/che/blob/master/dockerfiles/keycloak/kc_realm_user.sh#L24 Since we use templates directly in Keycloak entrypoint, all puppet syntax needs to be removed there.

@hkolvenbach will you add those changes to Keycloak script?

**Comment by dmytro-ndp** [comment]
But this problem is not specific for the pull request because there is another CI build which failed because of the same reasons: https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/175/consoleText

Theoretically, having last changes from master in this branch could help to solve the problem.

**Comment by dmytro-ndp** [comment]
ci-test-docker-multi

**Comment by garagatyi** [comment]
@hkolvenbach ok

**Comment by eivantsov** [comment]
@hkolvenbach are there any plans to finalize this PR? Do you need any assistance on some input from maintainers of this part of code?

**Comment by ex-q** [comment]
@skabashnyuk does that btw. mean there is currently no https support for eclipse che on docker? Meaning not on the editor side, not for keycloak and not for the previewurls?

**Comment by eivantsov** [comment]
yes, there's currently no https support on Docker infra out of the box

**Comment by fwchen** [comment]
any update?

**Comment by fwchen** [comment]
i think it can request wss when url is https( javascript side), and it will support nginx reverse mode.

**Comment by HuJK** [comment]
Hello, about this step:
copy che.key (private key) and che.crt (full chain) certificates for your *.domain.tld to instance/config/traefik/ within your Che installation

where is the Che installation? The Che should installed in docker, right?

**Pull request closed without merging**
